File: LinkedList.js
class ListNode {
    constructor(value, next) {
        this.data = value;
        this.next = !next ? null : next;
    }
}

class LinkedList {
    #head;
    constructor() {
        this.#head = null;
    }

    push(value) {
        const newNode = new ListNode(value);
        let temp = this.#head;
        if (!temp) {
            this.#head = newNode
            return;
        }

        while (temp.next) {
            temp = temp.next;
        }

        temp.next = newNode;
    } 

    unshift(value) {
        const newNode = new ListNode(value);
        newNode.next = this.#head;
        this.#head = newNode;
    }


    print() {
        let current = this.#head;
        if (!this.#head) return console.log('This is an empty list');
        while (current) {
            console.log(current.data);
            current = current.next;
        }
    }


    reverse() {
        let current = this.#head;
        let previous = null;
        let temp = null;

        while (current) {
            temp = current.next;
            current.next = previous;
            previous = current;
            current = temp;
        }

        this.#head = previous
    }
}


const list = new LinkedList();

list.push(1);
list.push(5);
list.push(2);
list.unshift(0);
list.reverse();
list.push(10);
list.print();

File: LinkedList.py
class Node:
    def __init__(self, value) -> None:
        self.data = value if (value) else None
        self.next = None
        
class LinkedList:
    def __init__(self) -> None:
        self.head = None
        
    def push(self, value):
        node = Node(value)
        if not self.head: 
            self.head = node
        else:
            temp = self.head
            while temp.next != None:
                temp = temp.next
            temp.next = node
            
    def print(self):
        current = self.head
        while current != None:
            print(f'{current.data}')
            current = current.next

    def unshift(self, value):
        node = Node(value)
        if not self.head: self.head = node
        else:
            node.next = self.head
            self.head = node
            
    def toList(self):
        list = []
        current = self.head
        while current != None:
            list.append(current.data)
            current = current.next
        return list
    
    def reverse(self):
        current = self.head
        previous = None
        temp = None
        while current != None:
            temp = current.next
            current.next = previous
            previous = current
            current = temp
        self.head = previous 
            
list = LinkedList()
for i in [10, 2, 6]: list.push(i)
list.unshift(1)
print(list.toList())
list.reverse()
print(list.toList())
